runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/observers/DirectStreamObserver.java

```diff
@@ -162,24 +170,23 @@ public void onNext(T value) throws StreamObserverCancelledException {
   public void onError(Throwable t) {
     isReadyNotifier.forceTermination();
     synchronized (lock) {
-      markClosedOrThrow();
-      outboundObserver.onError(t);
+      if (!isClosed) {
```

Comment:
I think we should have
isUserClosed = onError or onCompleted called on this object
isClosed = onError or onCompleted called on outboundObserver
Could name isClosed and isOutboundClosed or something if clearer. Woudl be good to comment above too.

I think that onError/onCompleted should be like:

check(!isUserClosed);
isUserClosed = true;
if (!isClosed) {
  outboundObserver.onError(t);
  isClosed = true; 
}

enforcing that only one of them is called. And terminate should be like

if (!isClosed) {
  outboundObserver.onError(t);
  isClosed = true;
}

Reply:
done
also caught the illegal state exception in `AbstractWindmillStream#halfClose`